File: packages/c/src/detokenize.c
/* Stateful detokenizer: byte_level + metaspace + byte fallback + partial
 * UTF-8 buffering across calls. Mirrors @codecai/web's Detokenizer. */
#include "codec/codec.h"
#include "codec_internal.h"

#include <stdlib.h>
#include <string.h>

/* Output is a growable byte buffer that we always finalize as UTF-8. */
typedef struct {
    char  *data;
    size_t len;
    size_t cap;
} growbuf_t;

static int gb_reserve(growbuf_t *b, size_t need) {
    if (b->cap >= need) return 1;
    size_t cap = b->cap ? b->cap : 64;
    while (cap < need) cap *= 2;
    char *p = (char *)realloc(b->data, cap + 1);
    if (!p) return 0;
    b->data = p;
    b->cap = cap;
    return 1;
}
static int gb_append(growbuf_t *b, const char *s, size_t len) {
    if (!gb_reserve(b, b->len + len)) return 0;
    memcpy(b->data + b->len, s, len);
    b->len += len;
    return 1;
}
static int gb_append_bytes(growbuf_t *b, const uint8_t *s, size_t len) {
    return gb_append(b, (const char *)s, len);
}
static int gb_append_replacement(growbuf_t *b) {
    /* U+FFFD = EF BF BD */
    static const char repl[3] = { (char)0xEF, (char)0xBF, (char)0xBD };
    return gb_append(b, repl, 3);
}

/* ── Detokenizer struct ─────────────────────────────────────────────────── */

struct codec_detokenizer {
    const codec_tokenizer_map_t *map;

    /* Partial-UTF-8 buffer; only populated by byte_level decoded bytes
     * and SentencePiece byte-fallback IDs. */
    uint8_t *byte_buf;
    size_t   byte_len;
    size_t   byte_cap;

    int32_t fallback_start; /* cached from map (-1 if none) */
    int32_t fallback_end;
    int     is_byte_level;
};

static int byte_buf_reserve(codec_detokenizer_t *d, size_t need) {
    if (d->byte_cap >= need) return 1;
    size_t cap = d->byte_cap ? d->byte_cap : 16;
    while (cap < need) cap *= 2;
    uint8_t *p = (uint8_t *)realloc(d->byte_buf, cap);
    if (!p) return 0;
    d->byte_buf = p;
    d->byte_cap = cap;
    return 1;
}

codec_status_t codec_detokenizer_new(const codec_tokenizer_map_t *map,
                                     codec_detokenizer_t **out) {
    if (!map || !out) return CODEC_ERR_INVALID_ARG;
    codec_detokenizer_t *d = (codec_detokenizer_t *)calloc(1, sizeof(*d));
    if (!d) return CODEC_ERR_OUT_OF_MEMORY;
    d->map = map;
    d->fallback_start = codec_map_byte_fallback_start(map);
    d->fallback_end   = codec_map_byte_fallback_end(map);
    d->is_byte_level  = codec_map_encoder(map) == CODEC_ENCODER_BYTE_LEVEL;
    *out = d;
    return CODEC_OK;
}

void codec_detokenizer_free(codec_detokenizer_t *d) {
    if (!d) return;
    free(d->byte_buf);
    free(d);
}

void codec_detokenizer_reset(codec_detokenizer_t *d) {
    if (d) d->byte_len = 0;
}

/* ── Byte-buffer flushing ───────────────────────────────────────────────── */

/* Try to peel and emit complete UTF-8 sequences from the head of byte_buf.
 * Stops when the next sequence is incomplete (returns leaving partial in
 * the buffer). */
static int flush_complete(codec_detokenizer_t *d, growbuf_t *out) {
    while (d->byte_len > 0) {
        int needed = codec_utf8_seq_len(d->byte_buf[0]);
        if (needed == 0) {
            /* Invalid leading byte: drop it as a replacement char. */
            memmove(d->byte_buf, d->byte_buf + 1, --d->byte_len);
            if (!gb_append_replacement(out)) return 0;
            continue;
        }
        if ((size_t)needed > d->byte_len) return 1; /* incomplete: keep buffered */

        /* Check continuation bytes. */
        int valid = 1;
        for (int k = 1; k < needed; k++) {
            if ((d->byte_buf[k] & 0xC0) != 0x80) { valid = 0; break; }
        }
        if (!valid) {
            /* Drop one byte, emit replacement. */
            memmove(d->byte_buf, d->byte_buf + 1, --d->byte_len);
            if (!gb_append_replacement(out)) return 0;
            continue;
        }
        if (!gb_append_bytes(out, d->byte_buf, (size_t)needed)) return 0;
        memmove(d->byte_buf, d->byte_buf + needed, d->byte_len - (size_t)needed);
        d->byte_len -= (size_t)needed;
    }
    return 1;
}

/* Force-flush whatever's in the byte buffer, replacing invalid sequences. */
static int flush_force(codec_detokenizer_t *d, growbuf_t *out) {
    /* Walk through byte_buf decoding; whatever doesn't decode → replacement. */
    while (d->byte_len > 0) {
        int needed = codec_utf8_seq_len(d->byte_buf[0]);
        if (needed == 0 || (size_t)needed > d->byte_len) {
            memmove(d->byte_buf, d->byte_buf + 1, --d->byte_len);
            if (!gb_append_replacement(out)) return 0;
            continue;
        }
        int valid = 1;
        for (int k = 1; k < needed; k++) {
            if ((d->byte_buf[k] & 0xC0) != 0x80) { valid = 0; break; }
        }
        if (!valid) {
            memmove(d->byte_buf, d->byte_buf + 1, --d->byte_len);
            if (!gb_append_replacement(out)) return 0;
            continue;
        }
        if (!gb_append_bytes(out, d->byte_buf, (size_t)needed)) return 0;
        memmove(d->byte_buf, d->byte_buf + needed, d->byte_len - (size_t)needed);
        d->byte_len -= (size_t)needed;
    }
    return 1;
}

/* ── Render ─────────────────────────────────────────────────────────────── */

codec_status_t codec_detokenizer_render(codec_detokenizer_t *d,
                                        const uint32_t *ids, size_t ids_len,
                                        codec_detokenize_opts_t opts,
                                        char **out, size_t *out_len) {
    if (!d || !out) return CODEC_ERR_INVALID_ARG;
    *out = NULL;
    if (out_len) *out_len = 0;

    growbuf_t buf = {0};
    if (!gb_reserve(&buf, 16)) return CODEC_ERR_OUT_OF_MEMORY;

    for (size_t i = 0; i < ids_len; i++) {
        uint32_t id = ids[i];

        /* Byte-fallback range: SentencePiece reserves IDs for bytes 0x00-0xFF. */
        if (d->fallback_start >= 0
            && (int64_t)id >= (int64_t)d->fallback_start
            && (int64_t)id <= (int64_t)d->fallback_end) {
            if (!byte_buf_reserve(d, d->byte_len + 1)) {
                free(buf.data); return CODEC_ERR_OUT_OF_MEMORY;
            }
            d->byte_buf[d->byte_len++] = (uint8_t)((int64_t)id - d->fallback_start);
            if (!flush_complete(d, &buf)) { free(buf.data); return CODEC_ERR_OUT_OF_MEMORY; }
            continue;
        }

        if (d->is_byte_level) {
            /* byte_level: every vocab token IS a byte sequence. */
            if (codec_map_is_special(d->map, id) && !opts.render_special) {
                if (d->byte_len > 0) {
                    if (!flush_force(d, &buf)) { free(buf.data); return CODEC_ERR_OUT_OF_MEMORY; }
                }
                continue;
            }
            const codec_id_entry_t *e = codec_map_entry(d->map, id);
            if (!e) {
                if (d->byte_len > 0) {
                    if (!flush_force(d, &buf)) { free(buf.data); return CODEC_ERR_OUT_OF_MEMORY; }
                }
                if (!gb_append_replacement(&buf)) { free(buf.data); return CODEC_ERR_OUT_OF_MEMORY; }
                continue;
            }
            if (!byte_buf_reserve(d, d->byte_len + e->len)) {
                free(buf.data); return CODEC_ERR_OUT_OF_MEMORY;
            }
            memcpy(d->byte_buf + d->byte_len, e->bytes, e->len);
            d->byte_len += e->len;
            if (!flush_complete(d, &buf)) { free(buf.data); return CODEC_ERR_OUT_OF_MEMORY; }
            continue;
        }

        /* metaspace / identity: token text rendered directly. */
        if (d->byte_len > 0) {
            if (!flush_force(d, &buf)) { free(buf.data); return CODEC_ERR_OUT_OF_MEMORY; }
        }
        if (codec_map_is_special(d->map, id) && !opts.render_special) continue;

        const codec_id_entry_t *e = codec_map_entry(d->map, id);
        if (!e) {
            if (!gb_append_replacement(&buf)) { free(buf.data); return CODEC_ERR_OUT_OF_MEMORY; }
            continue;
        }
        if (!gb_append_bytes(&buf, e->bytes, e->len)) {
            free(buf.data); return CODEC_ERR_OUT_OF_MEMORY;
        }
    }

    if (!opts.partial && d->byte_len > 0) {
        if (!flush_force(d, &buf)) { free(buf.data); return CODEC_ERR_OUT_OF_MEMORY; }
    }

    if (!gb_reserve(&buf, buf.len + 1)) { free(buf.data); return CODEC_ERR_OUT_OF_MEMORY; }
    buf.data[buf.len] = 0;
    *out = buf.data;
    if (out_len) *out_len = buf.len;
    return CODEC_OK;
}

File: packages/rust/src/detokenize.rs
// SPDX-License-Identifier: MIT
//! Stateful detokenizer: token IDs → text.
//!
//! Three correctness concerns it handles:
//!
//! 1. Per-token decoding via the map's encoder (byte_level / metaspace / identity).
//! 2. Byte-fallback range: IDs in `[byte_fallback_start, byte_fallback_end]` are decoded as raw bytes.
//! 3. Partial multi-byte sequences across frame boundaries: buffered between calls when `partial: true`.

use std::collections::{HashMap, HashSet};

use crate::byte_encoder::{decode_byte_level_token, METASPACE};
use crate::map::TokenizerMap;

/// Options for [`Detokenizer::render`].
#[derive(Debug, Clone, Copy, Default)]
pub struct DetokenizeOptions {
    /// If `true`, this is not the final chunk: buffer any trailing
    /// partial UTF-8 sequence rather than emitting replacement
    /// characters. Set to `false` on the last chunk so the buffer flushes.
    pub partial: bool,
    /// If `true`, render special tokens (e.g. `<|eos|>`) as text. Default: false.
    pub render_special: bool,
}

/// Stateful detokenizer.
///
/// Construct with a [`TokenizerMap`]; call [`Detokenizer::render`]
/// repeatedly with chunks of IDs. State (the partial UTF-8 byte buffer)
/// persists across calls; [`Detokenizer::reset`] clears it.
pub struct Detokenizer {
    special_ids: HashSet<u32>,
    fallback_start: i64,
    fallback_end: i64,
    /// `Some` when encoder == "byte_level".
    id_to_bytes: Option<HashMap<u32, Vec<u8>>>,
    /// `Some` for metaspace + identity.
    id_to_text: Option<HashMap<u32, String>>,
    byte_buffer: Vec<u8>,
}

impl Detokenizer {
    /// Build a detokenizer from a map.
    pub fn new(map: &TokenizerMap) -> Self {
        let special_ids: HashSet<u32> = map
            .special_tokens
            .as_ref()
            .map(|s| s.values().copied().collect())
            .unwrap_or_default();
        let fallback_start = map.byte_fallback_start.unwrap_or(-1);
        let fallback_end = map.byte_fallback_end.unwrap_or(-2);

        let (id_to_bytes, id_to_text) = if map.encoder.as_deref() == Some("byte_level") {
            (Some(build_byte_level_table(map)), None)
        } else {
            (None, Some(build_text_table(map)))
        };

        Self {
            special_ids,
            fallback_start,
            fallback_end,
            id_to_bytes,
            id_to_text,
            byte_buffer: Vec::new(),
        }
    }

    /// Render a chunk of IDs to text. Stateful across calls.
    pub fn render(&mut self, ids: &[u32], options: DetokenizeOptions) -> String {
        let mut out = String::new();
        let render_special = options.render_special;

        for &id in ids {
            // Byte-fallback range: SentencePiece reserves IDs for raw bytes 0x00-0xFF.
            let id_i = id as i64;
            if id_i >= self.fallback_start && id_i <= self.fallback_end {
                let b = (id_i - self.fallback_start) as u8;
                self.byte_buffer.push(b);
                self.flush_all_bytes(&mut out);
                continue;
            }

            if let Some(map_bytes) = &self.id_to_bytes {
                // byte_level: every vocab token IS a byte sequence.
                if self.special_ids.contains(&id) && !render_special {
                    if !self.byte_buffer.is_empty() {
                        self.flush_bytes_force(&mut out);
                    }
                    continue;
                }
                match map_bytes.get(&id) {
                    None => {
                        if !self.byte_buffer.is_empty() {
                            self.flush_bytes_force(&mut out);
                        }
                        out.push('\u{FFFD}');
                    }
                    Some(bytes) => {
                        self.byte_buffer.extend_from_slice(bytes);
                        self.flush_all_bytes(&mut out);
                    }
                }
                continue;
            }

            // metaspace / identity: token text is rendered directly.
            if !self.byte_buffer.is_empty() {
                self.flush_bytes_force(&mut out);
            }
            if self.special_ids.contains(&id) && !render_special {
                continue;
            }
            match self.id_to_text.as_ref().and_then(|m| m.get(&id)) {
                Some(text) => out.push_str(text),
                None => out.push('\u{FFFD}'),
            }
        }

        if !options.partial && !self.byte_buffer.is_empty() {
            self.flush_bytes_force(&mut out);
        }
        out
    }

    /// Reset internal state: call between conversations / requests.
    pub fn reset(&mut self) {
        self.byte_buffer.clear();
    }

    /// Convenience: detokenize a complete sequence in one shot. Uses a
    /// fresh detokenizer; partial buffering not exposed.
    pub fn detokenize(map: &TokenizerMap, ids: &[u32], render_special: bool) -> String {
        let mut d = Self::new(map);
        d.render(ids, DetokenizeOptions { partial: false, render_special })
    }

    // ── Internals ──────────────────────────────────────────────────────────

    fn flush_all_bytes(&mut self, out: &mut String) {
        loop {
            if self.byte_buffer.is_empty() {
                return;
            }
            let needed = utf8_sequence_length(self.byte_buffer[0]);
            if needed == 0 {
                self.byte_buffer.remove(0);
                out.push('\u{FFFD}');
                continue;
            }
            if self.byte_buffer.len() < needed {
                return;
            }
            let slice: Vec<u8> = self.byte_buffer.drain(..needed).collect();
            match std::str::from_utf8(&slice) {
                Ok(s) => out.push_str(s),
                Err(_) => out.push('\u{FFFD}'),
            }
        }
    }

    fn flush_bytes_force(&mut self, out: &mut String) {
        if self.byte_buffer.is_empty() {
            return;
        }
        let bytes = std::mem::take(&mut self.byte_buffer);
        // Lossy decode matches .NET's `Encoding.UTF8.GetString` (replacement char on invalid).
        out.push_str(&String::from_utf8_lossy(&bytes));
    }
}

fn utf8_sequence_length(b: u8) -> usize {
    if b & 0x80 == 0x00 {
        1
    } else if b & 0xE0 == 0xC0 {
        2
    } else if b & 0xF0 == 0xE0 {
        3
    } else if b & 0xF8 == 0xF0 {
        4
    } else {
        0
    }
}

fn build_byte_level_table(map: &TokenizerMap) -> HashMap<u32, Vec<u8>> {
    let mut result = HashMap::new();
    if let Some(vocab) = &map.vocab {
        result.reserve(vocab.len());
        for (token, &id) in vocab {
            result.insert(id, decode_byte_level_token(token));
        }
    }
    result
}

fn build_text_table(map: &TokenizerMap) -> HashMap<u32, String> {
    let mut result: HashMap<u32, String> = HashMap::new();
    let is_metaspace = map.encoder.as_deref() == Some("metaspace");

    if let Some(vocab) = &map.vocab {
        for (token, &id) in vocab {
            // SentencePiece byte-fallback tokens (<0xHH>) live in vocab
            // but are handled by the byte_fallback range path.
            if is_byte_fallback_token(token) {
                continue;
            }
            let text = if is_metaspace {
                token.replace(METASPACE, " ")
            } else {
                token.clone()
            };
            result.insert(id, text);
        }
    }
    if let Some(tokens) = &map.tokens {
        for (id_str, text) in tokens {
            if let Ok(id) = id_str.parse::<u32>() {
                result.insert(id, text.clone());
            }
        }
    }
    result
}

fn is_byte_fallback_token(s: &str) -> bool {
    let bytes = s.as_bytes();
    if bytes.len() != 6 {
        return false;
    }
    if bytes[0] != b'<' || bytes[1] != b'0' || bytes[2] != b'x' || bytes[5] != b'>' {
        return false;
    }
    is_hex_byte(bytes[3]) && is_hex_byte(bytes[4])
}

fn is_hex_byte(b: u8) -> bool {
    b.is_ascii_hexdigit()
}
